### PR TITLE
Project Tab

### PR DESCRIPTION
**Pull request by shiroyuki**
(resolves the indexer #289 and the webservice #57)

* The implementation is based on [Technical Considerations in Google Docs](https://docs.google.com/document/d/1GllbMBdS_BNOsQcepf-j4siYqK26v_DW0WAeKd0ny1g/edit#) (from #226) which is designed according to [v9.0.0 Project Schema](http://schema.staging.data.humancellatlas.org/type/project/9.0.0/project).
* The implementation is also backward-compatible.

### CODE REVIEW

**Comment by shiroyuki** [comment]
Just a friendly reminder: Please note that before we rebase and merge this PR, we will have to clean up the commit history and handle **commonmark** issue properly.

**Comment by mikebaumann** [comment]
My understanding from @hannes-ucsc is that the Project tab work will remain on a branch until he gets back from vacation and has a chance to review it before merging to `develop`.
Given the work to date has been done on `feature-project-tab-indexer` off of develop, I think this is the branch that will continue to be used until @hannes-ucsc returns.
Individual changes/PRs can be made off of `feature-project-tab-indexer` then merged back into `feature-project-tab-indexer`.

In short, I can review the changes to, date and those going forward,  and work can/should continue on this branch, but this branch should not be merged to `develop` until Hannes is back.

File: src/azul/project/hca/transformers.py [hunk 1]
```diff
@@ -168,15 +172,20 @@ def create_documents(self,
                          manifest: List[JSON],
                          metadata_files: Mapping[str, JSON]
                          ) -> Sequence[ElasticSearchDocument]:
+        # Create a bundle object.
```
Comment (mikebaumann) [inline src/azul/project/hca/transformers.py]: I think this is clear enough from the code and doesn't warrant adding a comment.

File: src/azul/project/hca/transformers.py [hunk 2]
```diff
@@ -168,15 +172,20 @@ def create_documents(self,
...
         bundle = api.Bundle(uuid=uuid,
                             version=version,
                             manifest=manifest,
                             metadata_files=metadata_files)
+
+        # Create ElasticSearch documents
         for file in bundle.files.values():
             visitor = TransformerVisitor()
+
```
Comment (mikebaumann) [inline src/azul/project/hca/transformers.py]: It looks like the difference in the blank line is because whitespace was present and is now removed, so the line does not contain unnecessary whitespace? If so, good.

File: src/azul/project/hca/transformers.py [hunk 3]
```diff
@@ -204,15 +210,20 @@ def create_documents(self,
                          manifest: List[JSON],
                          metadata_files: Mapping[str, JSON]
                          ) -> Sequence[ElasticSearchDocument]:
+        # Create a bundle object.
```
Comment (mikebaumann) [inline src/azul/project/hca/transformers.py]: Ditto comment above.

File: src/azul/project/hca/transformers.py [hunk 4]
```diff
@@ -224,3 +235,65 @@ def create_documents(self,
...
+
+        bundle_uuid = str(bundle.uuid)
+        simplified_project = _project_dict(bundle)
+
+        # Create ElasticSearch documents
+        for project in bundle.projects.values():
+            contents = dict(specimens=self._get_unique_list_of_specimens(bundle),
+                            files=self._get_unique_list_of_files(bundle),
+                            processes=self._get_unique_list_of_process(bundle),
+                            project=simplified_project)
```
Comment (mikebaumann) [inline src/azul/project/hca/transformers.py]: I am concerned that this implementation of `ProjectTransformer` departs from existing patterns and introduces code duplication, which could lead to inconsistencies in the future.

Instead, I would try implementing `ProjectTransformer` something like this:
```
class ProjectTransformer(Transformer):
    @property
    def entity_name(self):
        return 'projects'

    def create_documents(self,
                         uuid: str,
                         version: str,
                         manifest: List[JSON],
                         metadata_files: Mapping[str, JSON]
                         ) -> Sequence[ElasticSearchDocument]:
        # Create a bundle object.
        bundle = api.Bundle(uuid=uuid,
                            version=version,
                            manifest=manifest,
                            metadata_files=metadata_files)

        # Create ElasticSearch documents

        # Gather information on files in the bundle
        for file in bundle.files.values():
            files_visitor = TransformerVisitor()

            # Visit the relatives
            file.accept(files_visitor)  # Visit descendants
            file.ancestors(files_visitor)

        # Gather information on specimens in the bundle
        for specimen in bundle.specimens:
            specimen_visitor = TransformerVisitor()

            # Visit the relatives
            specimen.accept(specimen_visitor)  # Visit descendants
            specimen.ancestors(specimen_visitor)

        # Assign the contents to the ES doc
        contents = dict(specimens=_specimen_dict(specimen_visitor.specimens),
                        files=list(files_visitor.files.values()),
                        processes=list(specimen_visitor.processes.values()),
                        project=_project_dict(bundle))
        es_document = ElasticSearchDocument(entity_type=self.entity_name,
                                            entity_id=str(project.document_id),
                                            bundles=[Bundle(uuid=str(bundle.uuid),
                                                            version=bundle.version,
                                                            contents=contents)])
        yield es_document
```

I think this is close to something that would work, although, to be clear, I have not yet tested it as I still need to look at how the test infrastructure and tests are setup for this stuff. Yet, you get the idea.

File: src/azul/project/hca/transformers.py [hunk 5]
```diff
@@ -98,13 +102,16 @@ def __init__(self) -> None:
 
     def visit(self, entity: api.Entity) -> None:
         if isinstance(entity, api.SpecimenFromOrganism):
+            # This is to ensure that we have a unique set of specimens.
```
Comment (mikebaumann) [inline src/azul/project/hca/transformers.py]: I think these comments (`This is to ensure that we have a unique set of ...`) are unnecessary and should be removed. The primary purpose of these lines is to add the data to the visitor (and to do in a way that ensures uniqueness). This seems clear enough from the code.

Comment (shiroyuki) [reply src/azul/project/hca/transformers.py]: @mikebaumann, I think the comment is necessary.

As the way the code is designed, this method does not explicitly express the reason why we keep the data in a dictionary. It would require developers to look into the usage to get the idea why we need to collect the data in a dictionary, not in a list or set.

A few things I would prefer to have more explanation written/documented are:

* If the key is registered, why does the method always override the value (or the extracted version of the value)? Is it to simplify the code by not checking if the key already exists?
* Is there a chance that there are two slightly different value with the same key? If so, supposing that we choose to override the value for whatever reason, which version of value should we use?

I believe that in-code documentation will not help not just a newcomer, like myself, but other developers to understand what the code does and why it is designed to work this way.

File: src/azul/project/hca/transformers.py (hunk 4 above)
Comment (shiroyuki) [reply src/azul/project/hca/transformers.py]: As I mentioned in Slack, I agree with your suggestion and will work on the update.

File: src/azul/project/hca/transformers.py (hunk 3 above)
Comment (shiroyuki) [reply src/azul/project/hca/transformers.py]: Got it. Will remove this in the next commit.

File: src/azul/project/hca/transformers.py (hunk 1 above)
Comment (shiroyuki) [reply src/azul/project/hca/transformers.py]: Got it. Will remove this in the next commit.

File: src/azul/project/hca/transformers.py (hunk 2 above)
Comment (shiroyuki) [reply src/azul/project/hca/transformers.py]: 👍

File: src/azul/project/hca/transformers.py (hunk 4 above)
Comment (mikebaumann) [reply src/azul/project/hca/transformers.py]: Thanks, @shiroyuki, I reposted this from our DM Slack chat so @hannes-ucsc has visibility to the comment history.

Also, on closer examination, I think the above proposal can be implemented using a single instance of `TransformerVisitor`. Please let me know if you have questions. Thanks!

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 6]
```diff
@@ -331,7 +483,11 @@ def _merge(self, dict_1, dict_2, identifier):
                 merged_dict[key] = list(merged_dict[key] + [value])
             elif isinstance(value, list):
                 cleaned_list = list(filter(None, chain(value, merged_dict[key])))
-                merged_dict[key] = list(set(cleaned_list))
+                if len(cleaned_list) > 0 and isinstance(cleaned_list[0], dict):  # make dicts hashable
```
Comment (hannes-ucsc) [inline src/azul/service/responseobjects/hca_response_v5.py]: Not a fan of the placement of this comment. The line below freezes the dicts, not this one.

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 7]
```diff
@@ -331,7 +483,11 @@ def _merge(self, dict_1, dict_2, identifier):
...
+                    hashable_dicts = [freeze(d) for d in cleaned_list]
```
Comment (hannes-ucsc) [inline src/azul/service/responseobjects/hca_response_v5.py]: I think `hashable_dicts = freeze(cleaned_list)` should also work but the comment would have to change to `Make each dict in cleaned_list hashable`

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 8]
```diff
@@ -340,6 +496,11 @@ def _merge(self, dict_1, dict_2, identifier):
     def return_response(self):
         return self.apiResponse
 
+    @staticmethod
+    def to_camel_case(text: str):
```
Comment (hannes-ucsc) [inline src/azul/service/responseobjects/hca_response_v5.py]: Make this a function in an `azul.strings` module.

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 9]
```diff
@@ -390,9 +551,24 @@ def make_projects(self, entry):
             project.pop("_type")
             project_shortname = project["project_shortname"]
             translated_project = {
+                "projectTitle": project.get("project_title"),
                 "projectShortname": project_shortname,
-                "laboratory": list(set(project.get("laboratory")) if project.get("laboratory") else [])
```
Comment (hannes-ucsc) [inline src/azul/service/responseobjects/hca_response_v5.py]: :-)

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 10]
```diff
@@ -390,9 +551,24 @@ def make_projects(self, entry):
...
             }
+
+            if self.projects_response:
+                translated_project["projectDescription"] = project.get("project_description", [])
+                translated_project["contributors"] = project.get("contributors", [])
+                translated_project["publications"] = project.get("publications", [])
+
+                for contributor in translated_project['contributors']:
+                    for key in list(contributor.keys()):
+                        contributor[KeywordSearchResponse.to_camel_case(key)] = contributor.pop(key)
```
Comment (hannes-ucsc) [inline src/azul/service/responseobjects/hca_response_v5.py]: `KeywordSearchResponse.to_camel_case` is noisy. Hence the request to make it a function.

File: test/indexer/data/56a338fe-7554-4b5d-96a2-7df127a7640b.metadata.json [hunk 11]
```diff
@@ -194,89 +194,114 @@
...
-            "describedBy": "https://schema.humancellatlas.org/type/project/5.1.0/project",
+            "describedBy": "https://schema.humancellatlas.org/type/project/9.0.0/project",
             "project_core": {
                 "project_shortname": "Mouse Melanoma",
                 "project_description": "The cancer microenvironment is a complex ecosystem characterized by dynamic interactions between diverse cell types, including malignant, immune and stromal cells. Here, we performed single-cell RNA sequencing on CD45+ and CD45- cells isolated from tumour and lymph nodes during a mouse model of melanoma. The transcriptional profiles of these individual cells taken at different time points coupled with assembled T cell receptor sequences, allowed us to identify distinct immune subpopulations and delineate their developmental trajectory. Our study provides insights into the complex interplay among cells within the tumour microenvironment and presents a valuable resource for future translational applications.",
                 "project_title": "Melanoma infiltration of stromal and immune cells"
             },
-            "publications": [],
+            "publications": [
```
Comment (hannes-ucsc) [inline test/indexer/data/56a338fe-7554-4b5d-96a2-7df127a7640b.metadata.json]: Is this copied from a real bundle or did you write this from scratch?

File: test/indexer/data/aee55415-d128-4b30-9644-e6b2742fa32b.files.results.json [hunk 12]
```diff
@@ -123,12 +123,143 @@
...
                                     ],
+                                    "institutions": [
+                                        "DKFZ German Cancer Research Center",
+                                        "EMBL-EBI",
+                                        "University of Cambridge",
+                                        "University of Helsinki",
+                                        "Wellcome Trust Sanger Institute"
+                                    ],
+                                    "contact_names": [
+                                        "Angela,,Riedel",
```
Comment (hannes-ucsc) [inline test/indexer/data/aee55415-d128-4b30-9644-e6b2742fa32b.files.results.json]: What's up with the double commas?

File: test/indexer/data/aee55415-d128-4b30-9644-e6b2742fa32b.files.results.json [hunk 13]
```diff
@@ -123,12 +123,143 @@
                                     }
                                 ],
                                 "project": {
+                                    "project_title": "Melanoma infiltration of stromal and immune cells",
```
Comment (hannes-ucsc) [inline test/indexer/data/aee55415-d128-4b30-9644-e6b2742fa32b.files.results.json]: Is this file actually used?

File: test/indexer/test_hca_indexer.py [hunk 14]
```diff
@@ -193,14 +220,13 @@ def check_specimen_merge(es_results):
...
                     self.assertEqual(1, len(result_contents["files"]))
                     file_doc_ids.add(result_contents["files"][0]["uuid"])
-                elif result_dict["_index"] == config.es_index_name("specimens"):
+                elif (result_dict["_index"] == config.es_index_name("specimens") or
+                      result_dict["_index"] == config.es_index_name("projects")):
                     self.assertEqual(len(result_dict["_source"]["bundles"]), 2)
                     for bundle in result_dict["_source"]["bundles"]:
                         result_contents = bundle["contents"]
-                        # Each bundle in specimen list contains two files
+                        # Each bundle in should contain two files
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py]: Remove comment. It's obvious from the assertion.

File: test/service/__init__.py [hunk 15]
```diff
@@ -10,12 +10,13 @@ def lambda_name(cls) -> str:
         return "service"
 
     _data_loader = None
+    seed = None
```
Comment (hannes-ucsc) [inline test/service/__init__.py]: Add comment explaining what `seed` is or does.

File: test/service/test_repository_projects.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+
+from service import WebServiceTestCase
+
+
+class RepositoryProjectsEndpointTest(WebServiceTestCase):
+
+    seed = 123
+
+    @classmethod
+    def setUpClass(cls):
```
Comment (hannes-ucsc) [inline test/service/test_repository_projects.py]: This method doesn't do anything.

File: test/service/test_repository_projects.py [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+        """Set a seed so that we can test the detail response with a stable project ID"""
```
Comment (hannes-ucsc) [inline test/service/test_repository_projects.py]: This comment should go somewhere else.

File: test/service/test_response.py [hunk 18]
```diff
@@ -6,7 +6,8 @@
 
 import requests
 
-from azul.service.responseobjects.hca_response_v5 import KeywordSearchResponse, FileSearchResponse
+from azul.service.responseobjects.hca_response_v5 import (
```
Comment (hannes-ucsc) [inline test/service/test_response.py]: Aligned indent.

File: src/azul/project/hca/transformers.py [hunk 19]
```diff
@@ -208,11 +243,15 @@ def create_documents(self,
                             version=version,
                             manifest=manifest,
                             metadata_files=metadata_files)
+
```
[the remark this answers is not in the document]
Comment (shiroyuki) [reply src/azul/project/hca/transformers.py]: In my opinion, even though the method isn't that long, a single newline to separate between assignments, function calls, control structures, return statement etc. would make it quicker to find the code that we want. Here is an example.

```python
# The code is written to illustrate the readability discussion.
def str_concat(s1, s2):
  final_strings : List[str] = list()
  if not isinstance(s1, str):
    final_strings.append(str(s1))
  else:
    final_strings.append(s1)
  final_strings.append(str(s2) if not isinstance(s2, str) else s2)
  return ''.join(final_strings)
```

Now, with the same code with line separations:

```python
# The code is written to illustrate the readability discussion.
def str_concat(s1, s2):
  final_strings : List[str] = list()

  if not isinstance(s1, str):
    final_strings.append(str(s1))
  else:
    final_strings.append(s1)

  final_strings.append(str(s2) if not isinstance(s2, str) else s2)

  return ''.join(final_strings)
```

Even though the method in this example is not long at all, I believe the extra spaces would help developers navigate the code easier due to increase in readability (See [Active vs. passive](https://www.interaction-design.org/literature/article/the-power-of-white-space)) and make the coding style more consistent regardless to the length of the method or function.

@hannes-ucsc, since we don't have written coding guideline, what is your opinion on this?

File: test/indexer/data/aee55415-d128-4b30-9644-e6b2742fa32b.files.results.json (hunk 13 above)
Comment (Ben-Wu) [reply test/indexer/data/aee55415-d128-4b30-9644-e6b2742fa32b.files.results.json]: It's used in `test_index_correctness` in `test_hca_indexer.py`

File: test/indexer/data/56a338fe-7554-4b5d-96a2-7df127a7640b.metadata.json (hunk 11 above)
Comment (shiroyuki) [reply test/indexer/data/56a338fe-7554-4b5d-96a2-7df127a7640b.metadata.json]: @hannes-ucsc I made this from scratch. I use the real metadata of a random medical research publication from NIH, which is unrelated to the bundle.

File: test/indexer/data/aee55415-d128-4b30-9644-e6b2742fa32b.files.results.json (hunk 12 above)
Comment (hannes-ucsc) [reply test/indexer/data/aee55415-d128-4b30-9644-e6b2742fa32b.files.results.json]: Got it. Thank you.

File: src/azul/project/hca/transformers.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply src/azul/project/hca/transformers.py]: I prefer the former for the reason I mentioned. I am fully aware that this is a matter of preference but as the project lead I made a choice and everyone should stick with that decision.

File: src/azul/service/responseobjects/elastic_request_builder.py [hunk 20]
```diff
@@ -490,7 +495,9 @@ def transform_request(self,
                     for x in es_response_dict['hits']['hits']]
             # Create a KeywordSearchResponse object
             self.logger.info('Creating KeywordSearchResponse')
-            final_response = KeywordSearchResponse(hits)
+            final_response = KeywordSearchResponse(
+                hits,
+                projects_response=entity_type == 'projects')
```
[the remark this answers is not in the document]
Comment (Ben-Wu) [reply src/azul/service/responseobjects/elastic_request_builder.py]: This method would need to also be in a `ProjectFileSearchResponse` class so how could that be done cleanly?  I tried something in the latest commit (1f27936) with multiple inheritance like this:

```py
class ProjectFileSearchResponse(FileSearchResponse, ProjectKeywordSearchResponse):
    pass
```

But I'm not sure this is the best way to do it

**Comment by Ben-Wu** [comment]
@hannes-ucsc All comments should be addressed.  The only question I have is how to deal with `FileSearchResponse`/`ProjectFileSearchResponse`.  I made some changes but I don't think I have the best solution.

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 21]
```diff
@@ -288,32 +306,159 @@ def agg_contents(aggs_dict, agg_name, agg_form="buckets"):
...
+            kwargs['fileCount'] = self.agg_contents(self.aggregates, 'fileCount', agg_form='value')
+            kwargs['specimenCount'] = self.hits['total']
 
         self.apiResponse = SummaryRepresentation(**kwargs)
 
 
+class ProjectSummaryResponse(BaseSummaryResponse):
+    """Build summary field for each project in projects endpoint"""
+
+    @staticmethod
```
Comment (hannes-ucsc) [inline src/azul/service/responseobjects/hca_response_v5.py]: Incidentally static. Please convert to instance method.

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 22]
```diff
@@ -495,6 +652,38 @@ def __init__(self, hits):
         self.apiResponse = ApiResponse(**class_entries)
 
 
+class ProjectKeywordSearchResponse(KeywordSearchResponse):
+    def make_projects(self, entry):
```
Comment (hannes-ucsc) [inline src/azul/service/responseobjects/hca_response_v5.py]: This duplicates too much functionality from the overridden method. Take a look at the Template Method design pattern. I think that could be used to solve this problem.

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 23]
```diff
@@ -478,10 +628,17 @@ def map_entries(self, entry):
             bundles=self.make_bundles(entry)
         )
 
+    @staticmethod
+    def for_entity_type(hits, entity_type):
+        if entity_type == 'projects':
+            return ProjectKeywordSearchResponse(hits)
+        return ProjectKeywordSearchResponse(hits)
```
Comment (hannes-ucsc) [inline src/azul/service/responseobjects/hca_response_v5.py]: Why does this return ProjectKeywordSearchResponse in both cases? If that's a mistake, why don't the tests catch it?

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 24]
```diff
@@ -110,6 +112,14 @@ def create_object(cls, bucket):
         new_object.totalCellCountByOrgan = bucket['cell_count']['value']
         return new_object
 
+    @classmethod
+    def create_object_from_simple_count(cls, count):
+        new_object = cls()
```
Comment (hannes-ucsc) [inline src/azul/service/responseobjects/hca_response_v5.py]: Consider renaming `new_object` to `self`

File: src/azul/service/responseobjects/elastic_request_builder.py [hunk 25]
```diff
@@ -178,6 +179,10 @@ def create_request(
                 agg,
                 ElasticTransformDump.create_aggregate(
                     filters, facet_config, agg))
+
+        if entity_type == 'projects':  # Make project specific aggregations
```
Comment (hannes-ucsc) [inline src/azul/service/responseobjects/elastic_request_builder.py]: project-specific

File: src/azul/service/responseobjects/elastic_request_builder.py [hunk 26]
```diff
@@ -668,3 +674,47 @@ def transform_autocomplete_request(
         self.logger.info(
             "Returning the final response for transform_autocomplete_request")
         return final_response
+
+    @staticmethod
+    def create_project_summary_aggregates(es_search, request_config):
+        """Add per project aggregations to the request, specific to project response"""
```
Comment (hannes-ucsc) [inline src/azul/service/responseobjects/elastic_request_builder.py]: `per-project` and `"""` on separate line

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 27]
```diff
@@ -288,32 +306,159 @@ def agg_contents(aggs_dict, agg_name, agg_form="buckets"):
...
+            kwargs['specimenCount'] = self.agg_contents(self.aggregates, 'specimenCount', agg_form='value')
+        elif 'fileCount' in self.aggregates:
+            kwargs['fileCount'] = self.agg_contents(self.aggregates, 'fileCount', agg_form='value')
+            kwargs['specimenCount'] = self.hits['total']
 
         self.apiResponse = SummaryRepresentation(**kwargs)
 
 
+class ProjectSummaryResponse(BaseSummaryResponse):
+    """Build summary field for each project in projects endpoint"""
```
Comment (hannes-ucsc) [inline src/azul/service/responseobjects/hca_response_v5.py]: `"""` on separate line

File: test/indexer/data/56a338fe-7554-4b5d-96a2-7df127a7640b.metadata.json [hunk 28]
```diff
@@ -194,89 +194,114 @@
     },
     "project.json": {
         "content": {
-            "describedBy": "https://schema.humancellatlas.org/type/project/5.1.0/project",
+            "describedBy": "https://schema.humancellatlas.org/type/project/9.0.0/project",
```
Comment (hannes-ucsc) [inline test/indexer/data/56a338fe-7554-4b5d-96a2-7df127a7640b.metadata.json]: I can't find version 9.0.0 but there is version 9.0.2

Comment (shiroyuki) [reply test/indexer/data/56a338fe-7554-4b5d-96a2-7df127a7640b.metadata.json]: @hannes-ucsc It was in the staging two weeks ago.

File: src/azul/service/responseobjects/hca_response_v5.py (hunk 24 above)
Comment (Ben-Wu) [reply src/azul/service/responseobjects/hca_response_v5.py]: The other methods in this file all use `new_object`.  Should those be renamed too?

File: test/indexer/data/56a338fe-7554-4b5d-96a2-7df127a7640b.metadata.json (hunk 28 above)
Comment (hannes-ucsc) [reply test/indexer/data/56a338fe-7554-4b5d-96a2-7df127a7640b.metadata.json]: The link doesn't resolve. Please fix.

**Comment by Ben-Wu** [comment]
I made the requested changes including changing `KeywordSearchResponse` to reduce repeated code.  But at this point I strongly believe that passing in `entity_type` into the constructor is the cleanest and most maintainable way to handle the case for both `KeywordSearchResponse` and `FileSearchResponse`.

`final_response = KeywordSearchResponse(hits, entity_type)`

File: src/azul/project/hca/transformers.py [hunk 29]
```diff
@@ -171,6 +201,7 @@ def create_documents(self,
                             version=version,
                             manifest=manifest,
                             metadata_files=metadata_files)
+        # Create ElasticSearch documents
```
Comment (hannes-ucsc) [inline src/azul/project/hca/transformers.py]: `Elasticsearch` (lower case s), but honestly, the method is called `create_documents` what does this comment add?

File: src/azul/service/responseobjects/hca_response_v5.py (hunk 24 above)
Comment (hannes-ucsc) [reply src/azul/service/responseobjects/hca_response_v5.py]: Nope. Consistency trumps. Leave as is.

File: src/azul/project/hca/transformers.py [hunk 30]
```diff
@@ -13,12 +13,44 @@
...
     reject(additional_projects, "Azul can currently only handle a single project per bundle")
+
+    # Store lists of all values of each of these facets to allow facet filtering
+    # and term counting on the webservice
+    laboratories = set()
+    institutions = set()
+    contact_names = set()
+    publication_titles = set()
+
+    for contributor in project.contributors:
```
Comment (hannes-ucsc) [inline src/azul/project/hca/transformers.py]: It really looks like you are extracting too many metadata properties in the metadata-api PR. The API should only expose the properties that are actually used.

**Comment by hannes-ucsc** [comment]
Last but not least, did you try this out in your deployment and if so, can I inspect the ES projects index in that deployment?

**Comment by Ben-Wu** [comment]
@hannes-ucsc Yes we tested it in out deployment.  You should be able to look at the `azul_projects_jyo` index.

For bundle testing I didn't know they weren't validated.  We're talking about the bundles in `examples/bundles/public-beta/` in the `metadata-schema` repo right?  Would it be reasonable to add tests there to validate those?

**Comment by hannes-ucsc** [comment]
I'm talking about the canned bundles in this repo, not the ones in the metadata-schema repo.

So for example this modification by @shiroyuki that I commented on: https://github.com/DataBiosphere/azul/pull/293/files#r219936247

The canned bundles were taken out of a real DSS so we know they are real. How do I know that the manual edits that @shiroyuki made reflect reality, or are at least valid against the schema?

Sorry this review is taking so long. My days are busy and I don't get to do reviews until late in the day.

The two main issues holding this up are the fact that the project-related changes in the metadata-api don't distinguish between optional and required fields and that they add too many unused fields. The metadata-api is not supposed to be a complete representation of the metadata. It is only supposed to represent the essential metadata fields, the ones we use in Azul nothing else. You PR there adds two new classes with what looks like a complete set of all fields described in the schema. And it adds a new type of data structure: a set of dataclass instances. I've been careful to avoid those. I prefer to keep it that way. So if you really need that type of data structure, we should at least ensure that the semantics are solid. Having `unsafe_hash` is not exactly solid, hence the name. So I would prefer to remove the set of dataclass instances and replace it with set of strings or at least freezing the dataclass instances so they can safely be used in sets and dicts. The key here is to minimize the number of new classes and fields that this adds to the metadata API.

File: requirements.txt [hunk 31]
```diff
@@ -10,7 +10,7 @@ Faker==0.8.11
 future==0.16.0
 hca==4.4.1
 git+git://github.com/hannes-ucsc/jmespath.py@hannes-fix-cache-race#egg=jmespath
-git+git://github.com/HumanCellAtlas/metadata-api@release/1.0b4#egg=hca-metadata-api
+git+git://github.com/HumanCellAtlas/metadata-api@feature/benwu/20-projects-tab-fields#egg=hca-metadata-api
```
Comment (Ben-Wu) [inline requirements.txt]: **TODO:** Change this to release version when release with projects tab changes is made

**Comment by Ben-Wu** [comment]
I made this into three commits.  One for indexer, one for service, and one for the clean up of both.
@hannes-ucsc Let me know if this makes sense to do

File: src/azul/indexer.py [hunk 32]
```diff
@@ -1,7 +1,7 @@
 from abc import ABC
 from collections import defaultdict
 import logging
-from typing import Callable, List, Mapping, MutableMapping, Sequence, Union
+from typing import Callable, List, MutableMapping, Sequence, Union
```
Comment (hannes-ucsc) [inline src/azul/indexer.py]: Remove this change.

File: src/azul/indexer.py [hunk 33]
```diff
@@ -15,7 +15,7 @@
 
 log = logging.getLogger(__name__)
 
-DocumentsById = Mapping[str, ElasticSearchDocument]
+DocumentsById = MutableMapping[str, ElasticSearchDocument]
```
Comment (hannes-ucsc) [inline src/azul/indexer.py]: Remove this change.

File: src/azul/indexer.py [hunk 34]
```diff
@@ -185,11 +185,10 @@ def _merge_existing_docs(self, documents_by_id: DocumentsById) -> DocumentsById:
                 cur_doc = ElasticSearchDocument.from_index(cur_doc)
                 new_doc = merged_documents_by_id[cur_doc.document_id]
                 cur_doc.update_with(new_doc)
-                merged_documents_by_id[cur_doc.document_id] = cur_doc
-            return merged_documents_by_id
+                documents_by_id[cur_doc.document_id] = cur_doc
         else:
```
Comment (hannes-ucsc) [inline src/azul/indexer.py]: Remove this change.

File: src/azul/indexer.py [hunk 35]
```diff
@@ -185,11 +185,10 @@ def _merge_existing_docs(self, documents_by_id: DocumentsById) -> DocumentsById:
...
             log.info("No existing documents to merge with.")
-            return documents_by_id
+        return documents_by_id
```
Comment (hannes-ucsc) [inline src/azul/indexer.py]: Remove this change.
